collector-sqs/src/main/java/zipkin/collector/awssqs/AwsSqsSpanProcessor.java

```diff
@@ -0,0 +1,155 @@
+import com.amazonaws.services.sqs.model.Message;
+import com.amazonaws.services.sqs.model.MessageAttributeValue;
+import com.amazonaws.services.sqs.model.ReceiveMessageRequest;
+import com.amazonaws.services.sqs.model.ReceiveMessageResult;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import zipkin.Codec;
+import zipkin.Component;
+import zipkin.Span;
+import zipkin.collector.Collector;
+import zipkin.internal.Nullable;
+import zipkin.storage.Callback;
+
+
+final class AwsSqsSpanProcessor implements Closeable, Component {
+
+  private static final Logger log = LoggerFactory.getLogger(AwsSqsSpanProcessor.class);
+
+  private final AmazonSQSAsync client;
+  private final Collector collector;
+  private final String queueUrl;
+  private final int waitTimeSeconds;
+  private final AtomicBoolean closed = new AtomicBoolean(false);
+  private final AtomicReference<CheckResult> status = new AtomicReference<>(CheckResult.OK);
+
+  AwsSqsSpanProcessor(AmazonSQSAsync client, Collector collector, String queueUrl,
+      int waitTimeSeconds) {
+    this.client = client;
+    this.collector = collector;
+    this.queueUrl = queueUrl;
+    this.waitTimeSeconds = waitTimeSeconds;
+  }
+
+  @Override public CheckResult check() {
+    return status.get();
+  }
+
+  @Override public void close() throws IOException {
+    if (!closed.getAndSet(true)) {
+      client.shutdown();
+    }
+  }
+
+  AwsSqsSpanProcessor run() {
+    final ReceiveMessageRequest request = new ReceiveMessageRequest(queueUrl)
+        .withWaitTimeSeconds(waitTimeSeconds)
+        .withMessageAttributeNames("spans");
+
+    if (!closed.get()) {
+      client.receiveMessageAsync(request, new Handler(this));
+    }
+
+    return this;
+  }
+
+  private void process(final List<Message> messages) {
+    final List<DeleteMessageBatchRequestEntry> deletes = new LinkedList<>();
+    for (Message message : messages) {
+      MessageAttributeValue spanAttributes = message.getMessageAttributes().get("spans");
+      if (spanAttributes != null) {
+        final List<Span> spans = new LinkedList<>();
+        byte[] bytes;
+        switch (spanAttributes.getDataType()) {
+          case "Binary.JSON":
+            bytes = spanAttributes.getBinaryValue().array();
+            spans.addAll(Codec.JSON.readSpans(bytes));
+            break;
+          case "Binary.THRIFT":
+            bytes = spanAttributes.getBinaryValue().array();
+            spans.addAll(Codec.THRIFT.readSpans(bytes));
+            break;
+          default:
+            log.error(String.format("Unsupported codec %s", spanAttributes.getDataType()));
+            continue;
+        }
+
+        collector.accept(spans, new CollectorCallback(spans));
+
+        deletes.add(
```

Comment:
This should probably be refactored a bit.  I will change it so that messages are only deleted if the collector accepts them and noop on the error case so they will be retried.